crate/server/src/core/operations/certify.rs

```diff
@@ -12,30 +37,121 @@ use crate::{
     }
-    if request.certificate_request_value.is_some() {
-        kms_bail!(KmsError::InvalidRequest(
-            "Certificate Request Value not supported".to_string()
-        ))
-    }
-    if request.attributes.is_none() {
-        kms_bail!(KmsError::InvalidRequest(
-            "No attributes provided".to_string()
-        ))
+
+    let mut attributes = request.attributes.clone().ok_or(KmsError::InvalidRequest(
+        "Attributes specifying the issuer private key id are missing".to_string(),
+    ))?;
+
+    // Retrieve and update tags
+    let mut tags = remove_tags(&mut attributes).unwrap_or_default();
+    if !tags.is_empty() {
+        check_user_tags(&tags)?;
     }
 
-    trace!("Certify arguments OK");
+    // Get the issuer private key
+    let issuer_private_key_id =
+        issuer_private_key_id_from_attributes(&attributes)?.ok_or_else(|| {
+            KmsError::InvalidRequest(
+                "The private key of the issuer is not found in the attributes".to_string(),
+            )
+        })?;
+    let issuer_private_key = retrieve_object_for_operation(
+        &issuer_private_key_id,
+        ObjectOperationType::Certify,
+        kms,
+        user,
+        params,
+    )
+    .await?;
+    // Convert to an openssl PrivateKey
+    let issuer_pkey: PKey<Private> = kmip_private_key_to_openssl(&issuer_private_key.object)?;
+
+    //retrieve the certificate associated with the private key
+    let (issuer_certificate_owm, issuer_certificate) = retrieve_certificate_for_private_key(
+        &issuer_private_key.object,
+        ObjectOperationType::Get,
+        kms,
+        user,
+        params,
+    )
+    .await?;
+
+    // Create a new Asn1Time object for the current time
+    let now = Asn1Time::days_from_now(0).context("could not get a date in ASN.1")?;
+
+    // retrieve the number of days for the validity of the certificate
+    // the number of days cannot exceed that of the issuer certificate
+    let number_of_days = min(
+        issuer_certificate.not_after().diff(&now)?.days as usize,
+        number_of_days_from_attributes(&attributes)?.unwrap_or(3650),
+    );
+
+    // Parse the CSR
+    let csr_bytes = request
+        .certificate_request_value
+        .context("the certificate request must be provided")?;
+    let csr = match request.certificate_request_type {
+        Some(CertificateRequestType::PEM) => X509Req::from_pem(&csr_bytes),
+        Some(CertificateRequestType::PKCS10) => X509Req::from_der(&csr_bytes),
+        Some(CertificateRequestType::CRMF) => kms_bail!(KmsError::InvalidRequest(
+            "Certificate Request Type CRMF not supported".to_string()
+        )),
+        None => kms_bail!(KmsError::InvalidRequest(
+            "the certificate Request Type must be provided (PEM or PKCS10)".to_string()
+        )),
+    }?;
+
+    // Create an X509 struct with the desired certificate information.
+    let mut x509_builder = X509::builder().unwrap();
+    x509_builder.set_version(3)?;
```

Comment:
X509 version is 0-indexed in OpenSSL. Setting version 3 should be
```suggestion
    x509_builder.set_version(2)?;
```
but I think defining a constant like
```
pub const X509_VERSION3: i32 = 2;
```
in `crate/server/src/core/certificate/mod.rs` gives more readability in the code ?

Reply:
done in https://github.com/Cosmian/kms/commit/5eff99f9be8c2226cc47fb8644d88257554bf093